evaluators/rank/AveragePrecision/__init__.py

```diff
@@ -0,0 +1,38 @@
+from typing import Sequence, Any
+import numpy as np
+from jina.executors.evaluators.rank import BaseRankingEvaluator
+
+
+class AveragePrecisionEvaluator(BaseRankingEvaluator):
+    """A :class:`AveragePrecisionEvaluator` evaluates the Average Precision of the search.
+       https://en.wikipedia.org/wiki/Evaluation_measures_(information_retrieval)#Average_precision
+    """
+
+    @property
+    def metric(self):
+        return 'AveragePrecision'
+
+    def evaluate(self, actual: Sequence[Any], desired: Sequence[Any], *args, **kwargs) -> float:
+        """"
+        :param actual: the matched document identifiers from the request as matched by jina indexers and rankers
+        :param desired: the expected documents matches ids sorted as they are expected
+        :return the evaluation metric value for the request document
+        """
+        matches = []
+        for idx, doc_id in enumerate(desired):
+            matches.append(doc_id in actual[:idx+1])
+        matches = np.array(matches)
+        cumsum = np.cumsum(matches)
+        precision = cumsum / np.arange(1, len(matches) + 1)
+
+        # Does not match behaviour in
```

Comment:
Not sure if you want to call it a bug as it's up for definition.
This precision array basically contains all precision@x where x is the idx of the array.
In the array here precision@0 is set to 1.0, whereas in the precision evaluator precision@0 will return 0.
Precision here must be set to 1.0 as the integral will otherwise return other results.
Most notably the test case desired=[0, 1, 2, 3] actual=[0, 1, 2, 3] would no longer return 1.0.

Reply:
Ah okey nice explanation